internals/overlord/logstate/manager.go

```diff
@@ -15,26 +15,82 @@
 package logstate
 
 import (
+	"sync"
+
+	"github.com/canonical/pebble/internals/logger"
 	"github.com/canonical/pebble/internals/plan"
 	"github.com/canonical/pebble/internals/servicelog"
 )
 
-type LogManager struct{}
+type LogManager struct {
+	mu        sync.Mutex
+	gatherers map[string]*logGatherer
+	buffers   map[string]*servicelog.RingBuffer
+	plan      *plan.Plan
+}
 
 func NewLogManager() *LogManager {
-	return &LogManager{}
+	return &LogManager{
+		gatherers: map[string]*logGatherer{},
+		buffers:   map[string]*servicelog.RingBuffer{},
+	}
 }
 
-// PlanChanged is called by the service manager when the plan changes. We stop
-// all running forwarders, and start new forwarders based on the new plan.
+// PlanChanged is called by the service manager when the plan changes. We update the list of gatherers for each forwarder based on the new plan.
 func (m *LogManager) PlanChanged(pl *plan.Plan) {
-	// TODO: implement
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	// Create a map to hold gatherers for the new plan.
+	// Old gatherers will be moved over or deleted.
+	newGatherers := make(map[string]*logGatherer, len(pl.LogTargets))
+
+	for _, target := range pl.LogTargets {
+		gatherer := m.gatherers[target.Name]
+		if gatherer == nil {
+			// Create new gatherer
+			var err error
+			gatherer, err = newLogGatherer(target)
+			if err != nil {
+				logger.Noticef("Internal error: couldn't create gatherer: %v", err)
+				break
+			}
+
+			go gatherer.loop()
+			newGatherers[target.Name] = gatherer
+		} else {
+			// Copy over existing gatherer
+			newGatherers[target.Name] = gatherer
+			delete(m.gatherers, target.Name)
+		}
+
+		// Update iterators for gatherer
+		gatherer.planChanged(pl, m.buffers)
+	}
+
+	// Old gatherers for now-removed targets need to be shut down.
+	for _, gatherer := range m.gatherers {
+		gatherer.stop()
```

Comment:
I think if a gatherer is removed, it should be a fairly forceful remove, it shouldn't care about flushing clients, so maybe we pass a force argument into stop here.